runtime(python): highlight ellipsis literals

jparise opened the pull request:
The ellipsis literal (`...`) can be used in multiple contexts:

- Placeholders:     `class Foo: ...`
- Containers:       `Tuple[int, ...]`
- Assignments:      `x = ...`

This is a trickier pattern to match because we can't rely on keyword boundaries, so we instead look for exactly three dots (`...`).

This does mean that we will match the `...` portion of `x...x`, which isn't valid Python syntax, but I think that's an acceptable trade-off that avoids making this pattern much more complex.

Ref: https://docs.python.org/3/library/constants.html#Ellipsis

jparise committed:
runtime(python): highlight ellipsis literals

The ellipsis literal (`...`) can be used in multiple contexts:

- Placeholders:     `class Foo: ...`
- Containers:       `Tuple[int, ...]`
- Assignments:      `x = ...`

This is a trickier pattern to match because we can't rely on keyword
boundaries, so we instead look for exactly three dots (`...`).

This does mean that we will match the `...` portion of `x...x`, which
isn't valid Python syntax, but I think that's an acceptable trade-off
that avoids making this pattern much more complex.

Ref: https://docs.python.org/3/library/constants.html#Ellipsis

jparise committed:
runtime(python): exclude doctest ellipsis prefixes

dkearns commented on runtime/syntax/python.vim:
```diff
@@ -310,6 +310,8 @@ if !exists("python_no_builtin_highlight")
 	\ contains=ALLBUT,pythonBuiltin,pythonClass,pythonFunction,pythonType,pythonAsync
 	\ transparent
+  " the ellipsis literal `...` can be used in multiple syntactic contexts
+  syn match   pythonEllipsis	"\%(^\|[^.]\)\zs\.\.\.\ze\%([^.]\|$\)" display
```
This won't match if it follows another syntax match.  E.g. `type(...)` if the parens are syntax matches.  You might want to use a lookahead as well for symmetry but it's not required.

```suggestion
  syn match   pythonBuiltin "\.\@1<!.\.\.\ze\%([^.]\|$\)" display
```

jparise replied:
That's a very nice suggestion. Thanks! I applied it to the latest revision.

(It still needs to be `pythonEllipsis` due to my most recent change for doctest awareness.)

dkearns replied:
Sorry I didn't intend to suggest renaming the group.  I think that may have been the name in an earlier draft and it was changed underneath my other open comment window.

jparise committed:
runtime(python): add ellipsis highlighting tests

jparise committed:
runtime(python): improve pythonEllipsis pattern

Use a negative lookbehind and lookahead to ensure we only match our
three consecutive dots. This also eliminates the need to explicitly
check for the start and end of the line.

jparise committed:
runtime(python): support 'doctest: +ELLIPSIS'

Also expand the set of test cases to include Numpy indexing.
